# <center>ENEM - pt. 2
<hr>

## <center>Inicialização
<hr>

In [1]:
from tcc_funcoes import *

=================================== Módulo importado com sucesso ===================================



In [2]:
dc_dfs = abre_arquivo_pickle('dc_dfs_sem_outliers.pkl')
dc_preditoras = abre_arquivo_pickle('dc_preditoras.pkl')
dc_target = abre_arquivo_pickle('dc_target.pkl')
edicoes = abre_arquivo_pickle('edicoes.pkl')
dc_nome_salvar = abre_arquivo_pickle('dc_nome_salvar.pkl')

## <center>Tratando para começar alguns treinamentos
<hr>

In [3]:
### Faixa Etária

dc_fx_etaria = {
    '<_17': 1,
    '17': 2,
    '18': 3,
    '19': 4,
    '20': 5,
    '21': 6,
    '22': 7,
    '23': 8,
    '24': 9,
    '25': 10, 
    '26_30': 11,
    '31_35': 12,
    '36_40': 13,
    '41_45': 14,
    '46_50': 15,
    '51_55': 16,
    '56_60': 17,
    '61_65': 18,
    '66_70': 19,
    '>_70': 20
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['fx_etaria_num'] = dc_dfs[nome]['fx_etaria'].map(dc_fx_etaria).astype('category')

del dc_fx_etaria
gc.collect()

0

In [4]:
### Sexo

dc_sexo = {
    'F': 1,
    'M': 2
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['sexo_num'] = dc_dfs[nome]['sexo'].map(dc_sexo).astype('category')

del dc_sexo
gc.collect()

0

In [5]:
### Estado Civil

dc_estado_civil = {
    'solteiro': 1,
    'casado': 2,
    'separado': 3,
    'viuvo': 4
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['estado_civil_num'] = dc_dfs[nome]['estado_civil'].map(dc_estado_civil).astype('category')

del dc_estado_civil
gc.collect()

0

In [6]:
### Cor/ Raça

dc_cor_raca = {
    'branca': 1,
    'preta': 2,
    'parda': 3,
    'amarela': 4,
    'indigena': 5
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['cor_raca_num'] = dc_dfs[nome]['cor_raca'].map(dc_cor_raca).astype('category')

del dc_cor_raca
gc.collect()

0

In [7]:
### Nacionalidade

dc_nacionalidade = {
    'brasileiro': 1,
    'naturalizado': 2,
    'estrangeiro': 3,
    'brasileiro_nascido_exterior': 4
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['nacionalidade_num'] = dc_dfs[nome]['nacionalidade'].map(dc_nacionalidade).astype('category')

del dc_nacionalidade
gc.collect()

0

In [8]:
### Status do Ensino Médio

dc_conclusao_ensino_medio = {
    'ja_terminou_ensino': 1,
    'termina_no_ano_da_prova': 2,
    'termina_apos_ano_da_prova': 3,
    'nao_concluiu_nem_cursa': 4
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['conclusao_ensino_medio_num'] = dc_dfs[nome]['conclusao_ensino_medio'].map(dc_conclusao_ensino_medio).astype('category')

del dc_conclusao_ensino_medio
gc.collect()

0

In [9]:
### Flag Treineiro

dc_treineiro = {
    'treineiro': 1,
    'nao_treineiro': 2
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['flag_treineiro_num'] = dc_dfs[nome]['flag_treineiro'].map(dc_treineiro).astype('category')

del dc_treineiro
gc.collect()

0

In [10]:
### Língua Estrangeira

dc_lingua_estrangeira = {
    'ingles': 0,
    'espanhol': 1
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['lingua_estrangeira_num'] = dc_dfs[nome]['lingua_estrangeira'].map(dc_lingua_estrangeira).astype('category')

del dc_lingua_estrangeira
gc.collect()

0

In [11]:
### Escolaridade dos Pai e da Mãe

dc_escolaridade = {
    'sem_estudo': 1,
    'nao_completou_4_ano_fundamental': 2,
    'nao_completou_8_ano_fundamental': 3,
    'nao_completou_ensino_medio': 4,
    'completou_ensino_medio': 5,
    'completou_superior': 6,
    'completou_pos_graduacao': 7
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['01_escolaridade_pai_num'] = dc_dfs[nome]['01_escolaridade_pai'].map(dc_escolaridade).astype('category')
    dc_dfs[nome]['02_escolaridade_mae_num'] = dc_dfs[nome]['02_escolaridade_mae'].map(dc_escolaridade).astype('category')

del dc_escolaridade
gc.collect()

0

In [12]:
### Ocupação dos Pai e da Mãe

dc_ocupacao = {
    'trabalhador_rural': 1,
    'servicos_geral_atendimento': 2,
    'operario_trabalhador_manual': 3,
    'tecnicos_pequenos_proprietarios': 4,
    'alta_qualificacao_gestao': 5
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['03_ocupacao_pai_num'] = dc_dfs[nome]['03_ocupacao_pai'].map(dc_ocupacao).astype('category')
    dc_dfs[nome]['04_ocupacao_mae_num'] = dc_dfs[nome]['04_ocupacao_mae'].map(dc_ocupacao).astype('category')

del dc_ocupacao
gc.collect()

0

In [13]:
### Quantidade de moradores

### Sem alterações necessárias

print(sorted(dc_dfs['Ciências Humanas']['05_qtde_moradores'].unique()))

for nome in dc_dfs.keys():

    dc_dfs[nome]['05_qtde_moradores_num'] = dc_dfs[nome]['05_qtde_moradores'].astype('category')

[np.float32(1.0), np.float32(2.0), np.float32(3.0), np.float32(4.0), np.float32(5.0), np.float32(6.0), np.float32(7.0), np.float32(8.0), np.float32(9.0), np.float32(10.0), np.float32(11.0), np.float32(12.0), np.float32(13.0), np.float32(14.0), np.float32(15.0), np.float32(16.0), np.float32(17.0), np.float32(18.0), np.float32(19.0), np.float32(20.0)]


In [14]:
### Renda familiar

dc_renda_familiar = {
    'sem_renda': 1,
    'ate_1.0': 2,
    'ate_1.5': 3,
    'ate_2.0': 4,
    'ate_2.5': 5,
    'ate_3.0': 6,
    'ate_4.0': 7,
    'ate_5.0': 8,
    'ate_6.0': 9,
    'ate_7.0': 10,
    'ate_8.0': 11,
    'ate_9.0': 12,
    'ate_10.0': 13,
    'ate_12.0': 14,
    'ate_15.0': 15,
    'ate_20.0': 16,
    'mais_20.0': 17
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['06_renda_familiar_num'] = dc_dfs[nome]['06_renda_familiar'].map(dc_renda_familiar).astype('category')

del dc_renda_familiar
gc.collect()

0

In [15]:
### Frequência de dias com trabalhador doméstico

dc_dias_trabalhador_domestico = {
    'nenhum_dia': 1,
    'ate_dois_dias': 2,
    'ate_quatro_dias': 3,
    'mais_cinco_dias': 4
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['07_dias_trabalhador_domestico_num'] = dc_dfs[nome]['07_dias_trabalhador_domestico'].map(dc_dias_trabalhador_domestico).astype('category')

del dc_dias_trabalhador_domestico
gc.collect()

0

In [16]:
### Quantidade de banheiros / quartos / carros / motocicletas / geladeiras / freezers / máquinas de lavar roupas / máquinas de secar roupas / micro-ondas / televisores em cor / celulares / computadores

### Isso são as questões 8 a 17, 19, 22 e 24

dc_qtde = {
    'zero': 1,
    'um': 2,
    'dois': 3,
    'tres': 4,
    'quatro_ou_mais': 5
}

### Flag de posse de aspirador de pó / aparelho de DVD / TV por assinatura / telefone fixo / internet

### Isso são as questões 18, 20, 21, 23 e 25

dc_posse = {
    'nao_possui': 1,
    'possui': 2
}

for nome in dc_dfs.keys():

    dc_dfs[nome]['08_qtde_banheiro_num'] = dc_dfs[nome]['08_qtde_banheiro'].map(dc_qtde).astype('category')
    dc_dfs[nome]['09_qtde_quarto_num'] = dc_dfs[nome]['09_qtde_quarto'].map(dc_qtde).astype('category')
    dc_dfs[nome]['10_qtde_carro_num'] = dc_dfs[nome]['10_qtde_carro'].map(dc_qtde).astype('category')
    dc_dfs[nome]['11_qtde_motocicleta_num'] = dc_dfs[nome]['11_qtde_motocicleta'].map(dc_qtde).astype('category')
    dc_dfs[nome]['12_qtde_geladeira_num'] = dc_dfs[nome]['12_qtde_geladeira'].map(dc_qtde).astype('category')
    dc_dfs[nome]['13_qtde_freezer_num'] = dc_dfs[nome]['13_qtde_freezer'].map(dc_qtde).astype('category')
    dc_dfs[nome]['14_qtde_maq_lavar_roupa_num'] = dc_dfs[nome]['14_qtde_maq_lavar_roupa'].map(dc_qtde).astype('category')
    dc_dfs[nome]['15_qtde_maq_secar_roupa_num'] = dc_dfs[nome]['15_qtde_maq_secar_roupa'].map(dc_qtde).astype('category')
    dc_dfs[nome]['16_qtde_micro_ondas_num'] = dc_dfs[nome]['16_qtde_micro_ondas'].map(dc_qtde).astype('category')
    dc_dfs[nome]['17_qtde_maq_lavar_louca_num'] = dc_dfs[nome]['17_qtde_maq_lavar_louca'].map(dc_qtde).astype('category')

    dc_dfs[nome]['18_flag_aspirador_po_num'] = dc_dfs[nome]['18_flag_aspirador_po'].map(dc_posse).astype('category')
    
    dc_dfs[nome]['19_qtde_tv_num'] = dc_dfs[nome]['19_qtde_tv'].map(dc_qtde).astype('category')
    
    dc_dfs[nome]['20_flag_aparelho_dvd_num'] = dc_dfs[nome]['20_flag_aparelho_dvd'].map(dc_posse).astype('category')
    dc_dfs[nome]['21_flag_tv_assinatura_num'] = dc_dfs[nome]['21_flag_tv_assinatura'].map(dc_posse).astype('category')
    
    dc_dfs[nome]['22_qtde_celular_num'] = dc_dfs[nome]['22_qtde_celular'].map(dc_qtde).astype('category')
    
    dc_dfs[nome]['23_flag_telefone_fixo_num'] = dc_dfs[nome]['23_flag_telefone_fixo'].map(dc_posse).astype('category')
    
    dc_dfs[nome]['24_qtde_computadores_num'] = dc_dfs[nome]['24_qtde_computadores'].map(dc_qtde).astype('category')
    
    dc_dfs[nome]['25_flag_internet_num'] = dc_dfs[nome]['25_flag_internet'].map(dc_posse).astype('category')

del dc_qtde, dc_posse
gc.collect()

0

In [17]:
dc_preditoras_num = {}

for nome in dc_dfs.keys():

    dc_preditoras_num[nome] = [col + '_num' for col in dc_preditoras[nome]]

del nome
gc.collect()

0

## <center>Random Forest para gerar feature importance
<hr>

In [18]:
# %%time

# import cuml

# dc_rf = {}
# df_feature_importance = pd.DataFrame(dc_preditoras_num['Ciências Humanas'], columns=['variável'])

# for nome, df in dc_dfs.items():

#     print(f'Random Forest - {nome}')

#     target = dc_target[nome]
#     preditoras = dc_preditoras_num[nome]

#     x = df[preditoras]
#     y = df[target]

#     modelo = cuml.ensemble.RandomForestRegressor()

#     modelo.fit(x, y)

#     dc_rf[nome] = modelo

#     v_importancias = zip(dc_preditoras_num[nome], 100 * modelo.feature_importances_)

#     df_feature_importance = df_feature_importance.merge(pd.DataFrame(v_importancias, columns=['variável', f'{nome}']), on='variável', how='left')

#     del x, y, modelo, target, preditoras, nome, v_importancias, df

# gc.collect()

# salva_arquivo_pickle('dc_rf_cuml.pkl', dc_rf)
# salva_arquivo_pickle('df_feature_importance.pkl', df_feature_importance)

In [19]:
# dc_rf = abre_arquivo_pickle('dc_rf_cuml.pkl')
# df_feature_importance = abre_arquivo_pickle('df_feature_importance.pkl')

## <center>Concentração de variáveis preditoras
<hr>

In [20]:
dc_concentracao = {}

for nome, df in dc_dfs.items():

    target = dc_target[nome]

    concentracao = {}

    for coluna in dc_preditoras_num[nome]:

        contagem = df[coluna].value_counts(normalize=True)

        concentracao[coluna] = contagem.max()

    dc_concentracao[nome] = pd.DataFrame.from_dict(concentracao, orient='index', columns=['concentracao']).sort_values(by='concentracao', ascending=False).reset_index().rename(columns={'index': 'variavel'})

del nome, df, target, coluna, contagem, concentracao
gc.collect()

0

In [21]:
for nome in dc_dfs.keys():

    print(f' {nome} '.center(60, '='))

    display(dc_concentracao[nome].query('concentracao >= 0.95'))

===================== Ciências Humanas =====================


variavel  concentracao
0            nacionalidade_num      0.977623
1  17_qtde_maq_lavar_louca_num      0.962113

==================== Ciências Natureza =====================


variavel  concentracao
0            nacionalidade_num      0.977758
1  17_qtde_maq_lavar_louca_num      0.962400

==================== Linguagem e Código ====================


variavel  concentracao
0            nacionalidade_num      0.977658
1  17_qtde_maq_lavar_louca_num      0.962016

======================== Matemática ========================


variavel  concentracao
0            nacionalidade_num      0.977803
1  17_qtde_maq_lavar_louca_num      0.962029

========================= Redação ==========================


variavel  concentracao
0            nacionalidade_num      0.977930
1  17_qtde_maq_lavar_louca_num      0.961394

Como essas variáveis possuem uma concentração muito alta, podemos eliminá-las do modelo para evitar problemas.

In [84]:
dc_remover_preditoras = {}

for nome in dc_dfs.keys():

    dc_remover_preditoras[nome] = pd.DataFrame(columns=['variavel', 'motivo'])

    dc_remover_preditoras[nome] = pd.concat([dc_remover_preditoras[nome], pd.DataFrame([{'variavel': 'nacionalidade_num', 'motivo': 'concentracao'}])], ignore_index=True)
    dc_remover_preditoras[nome] = pd.concat([dc_remover_preditoras[nome], pd.DataFrame([{'variavel': '17_qtde_maq_lavar_louca_num', 'motivo': 'concentracao'}])], ignore_index=True)

    del nome
    gc.collect()

## <center>Matriz de correlação Phik
<hr>

In [23]:
from phik import phik_matrix
from phik.report import plot_correlation_matrix

In [ ]:
# dc_phik = {}

# for nome, df in dc_dfs.items():

#     interval_cols = [dc_target[nome]]

#     corr = phik_matrix(df[dc_preditoras_num[nome] + [dc_target[nome]]], interval_cols=interval_cols)

#     corr_2 = pd.DataFrame(corr.unstack())

#     dc_phik[nome] = corr_2.reset_index().rename(columns={0: 'phik', 'level_0': 'var_1', 'level_1': 'var_2'})

#     print(f'Matriz de correlação Phik - {nome}')

#     del nome, df, interval_cols, corr, corr_2
#     gc.collect()

# salva_arquivo_pickle(f'dc_phik.pkl', dc_phik)

In [24]:
dc_phik = abre_arquivo_pickle(f'dc_phik.pkl')

In [42]:
for nome in dc_dfs.keys():

    df = dc_phik[nome]

    v_um = df.query("phik == 1 and var_1 != var_2")[['var_1', 'var_2']]

    print(f' {nome} '.center(60, '='))
    print()

    print(v_um)
    print()

    del df, v_um
    gc.collect()

===================== Ciências Humanas =====================

                          var_1                       var_2
176  conclusao_ensino_medio_num          flag_treineiro_num
209          flag_treineiro_num  conclusao_ensino_medio_num

==================== Ciências Natureza =====================

                          var_1                       var_2
176  conclusao_ensino_medio_num          flag_treineiro_num
209          flag_treineiro_num  conclusao_ensino_medio_num

==================== Linguagem e Código ====================

                          var_1                       var_2
176  conclusao_ensino_medio_num          flag_treineiro_num
209          flag_treineiro_num  conclusao_ensino_medio_num

======================== Matemática ========================

                          var_1                       var_2
176  conclusao_ensino_medio_num          flag_treineiro_num
209          flag_treineiro_num  conclusao_ensino_medio_num

========================= Re

Como as variáveis são perfeitamente correlacionadas, podemos eliminar algumas delas para evitar multicolinearidade.

In [85]:
for nome in dc_dfs.keys():

    # dc_remover_preditoras[nome] += ['conclusao_ensino_medio', 'flag_treineiro']

    dc_remover_preditoras[nome] = pd.concat([dc_remover_preditoras[nome], pd.DataFrame([{'variavel': 'conclusao_ensino_medio_num', 'motivo': 'correlacao entre preditoras'}])], ignore_index=True)
    dc_remover_preditoras[nome] = pd.concat([dc_remover_preditoras[nome], pd.DataFrame([{'variavel': 'flag_treineiro_num', 'motivo': 'correlacao entre preditoras'}])], ignore_index=True)

    del nome
    gc.collect()

In [86]:
for nome in dc_dfs.keys():

    print(f' {nome} '.center(60, '='))
    display(dc_remover_preditoras[nome])


===================== Ciências Humanas =====================


variavel                       motivo
0            nacionalidade_num                 concentracao
1  17_qtde_maq_lavar_louca_num                 concentracao
2   conclusao_ensino_medio_num  correlacao entre preditoras
3           flag_treineiro_num  correlacao entre preditoras

==================== Ciências Natureza =====================


variavel                       motivo
0            nacionalidade_num                 concentracao
1  17_qtde_maq_lavar_louca_num                 concentracao
2   conclusao_ensino_medio_num  correlacao entre preditoras
3           flag_treineiro_num  correlacao entre preditoras

==================== Linguagem e Código ====================


variavel                       motivo
0            nacionalidade_num                 concentracao
1  17_qtde_maq_lavar_louca_num                 concentracao
2   conclusao_ensino_medio_num  correlacao entre preditoras
3           flag_treineiro_num  correlacao entre preditoras

======================== Matemática ========================


variavel                       motivo
0            nacionalidade_num                 concentracao
1  17_qtde_maq_lavar_louca_num                 concentracao
2   conclusao_ensino_medio_num  correlacao entre preditoras
3           flag_treineiro_num  correlacao entre preditoras

========================= Redação ==========================


variavel                       motivo
0            nacionalidade_num                 concentracao
1  17_qtde_maq_lavar_louca_num                 concentracao
2   conclusao_ensino_medio_num  correlacao entre preditoras
3           flag_treineiro_num  correlacao entre preditoras

## <center>Avaliando as variáveis selecionados para serem retiradas
<hr>

In [98]:
for nome, df in dc_remover_preditoras.items():

    target = dc_target[nome]

    variaveis = df['variavel'].to_list()

    phik = dc_phik[nome].query("var_1 in @variaveis and var_2 in @target")

    concentracao = dc_concentracao[nome].query("variavel in @variaveis")

    df_analise = db.sql("""

        select

            a.*,

            b.concentracao,

            c.phik as phik_nota

        from
            df as a

        left join
            concentracao as b
        on
            a.variavel = b.variavel

        left join
            phik as c
        on
            a.variavel = c.var_1

    """).df().sort_values(by='motivo', ascending=False)

    print(f' {nome} '.center(60, '='))
    display(df_analise)

===================== Ciências Humanas =====================


variavel                       motivo  concentracao  \
2   conclusao_ensino_medio_num  correlacao entre preditoras      0.458549   
3           flag_treineiro_num  correlacao entre preditoras      0.822738   
0            nacionalidade_num                 concentracao      0.977623   
1  17_qtde_maq_lavar_louca_num                 concentracao      0.962113   

   phik_nota  
2   0.107189  
3   0.095049  
0   0.065547  
1   0.160541

==================== Ciências Natureza =====================


variavel                       motivo  concentracao  \
2   conclusao_ensino_medio_num  correlacao entre preditoras      0.456505   
3           flag_treineiro_num  correlacao entre preditoras      0.824190   
0            nacionalidade_num                 concentracao      0.977758   
1  17_qtde_maq_lavar_louca_num                 concentracao      0.962400   

   phik_nota  
2   0.085663  
3   0.079141  
0   0.059367  
1   0.179161

==================== Linguagem e Código ====================


variavel                       motivo  concentracao  \
2   conclusao_ensino_medio_num  correlacao entre preditoras      0.458819   
3           flag_treineiro_num  correlacao entre preditoras      0.822752   
0            nacionalidade_num                 concentracao      0.977658   
1  17_qtde_maq_lavar_louca_num                 concentracao      0.962016   

   phik_nota  
2   0.077987  
3   0.070694  
0   0.065346  
1   0.159698

======================== Matemática ========================


variavel                       motivo  concentracao  \
2   conclusao_ensino_medio_num  correlacao entre preditoras      0.457660   
3           flag_treineiro_num  correlacao entre preditoras      0.824724   
0            nacionalidade_num                 concentracao      0.977803   
1  17_qtde_maq_lavar_louca_num                 concentracao      0.962029   

   phik_nota  
2   0.074366  
3   0.082523  
0   0.064959  
1   0.201304

========================= Redação ==========================


variavel                       motivo  concentracao  \
2   conclusao_ensino_medio_num  correlacao entre preditoras      0.462576   
3           flag_treineiro_num  correlacao entre preditoras      0.824377   
0            nacionalidade_num                 concentracao      0.977930   
1  17_qtde_maq_lavar_louca_num                 concentracao      0.961394   

   phik_nota  
2   0.096100  
3   0.095237  
0   0.043548  
1   0.124726

A variável de nacionalidade apresenta uma concentração muito alta e apresenta uma correlação com a variável resposta (respecitivamente para cada variável resposta) muito baixa (entre 4.3548% e 6.5547%). Assim, iremos sim retirar do modelo.

In [101]:
for nome in dc_dfs.keys():

    dc_preditoras[nome] = dc_preditoras[nome].remove('nacionalidade')
    dc_preditoras_num[nome] = dc_preditoras_num[nome].remove('nacionalidade_num')

Iremos começar avaliando a correção entra as variáveis preditoras e a variável resposta. Para isso, vamos usar a matriz de correlação Phik.

Vamos avaliar se variáveis com concentrações altas de alguma categoria possuem correlação com a variável resposta.

In [ ]:
dc_concentracao = {}

for nome, df in dc_dfs.items():

    target = dc_target[nome]

    concentracao = {}

    for coluna in dc_preditoras[nome]:

        contagem = df[coluna].value_counts(normalize=True)

        if contagem.max() >= 0.85:

            concentracao[coluna] = contagem.max()

    dc_concentracao[nome] = pd.DataFrame.from_dict(concentracao, orient='index', columns=['concentracao']).sort_values(by='concentracao', ascending=False).reset_index().rename(columns={'index': 'variavel'})

del nome, df, target, coluna, contagem, concentracao
gc.collect()

In [ ]:
for nome, df in dc_matriz_phik.items():

    target = dc_target[nome]

    vars = dc_concentracao[nome]['variavel'].tolist()

    df = df.query(f"var_1 == '{target}' and var_2 in {vars}").reset_index(drop=True)

    print(f" {nome} ".center(60, '='))
    print()

    print(df[['var_2', 'valor_phik']])
    print()

del nome, df, target, vars
gc.collect()

Como era de se esperar, não há correlações significativas entre as variáveis com alta concentração e a variável resposta. Assim, essas variáveis serão as primeiras a serem removidas do modelo, através da técnica de eliminação recursiva de variáveis (RFE).

In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import OrdinalEncoder

df_humanas = dc_dfs['Ciências Humanas'].copy()
cols_preditoras = dc_preditoras['Ciências Humanas']
cat_cols = df_humanas[cols_preditoras].select_dtypes(include=['category', 'object']).columns

encoder = OrdinalEncoder()
df_humanas[cat_cols] = encoder.fit_transform(df_humanas[cat_cols])

# arvore = DecisionTreeRegressor(random_state=42)

# arvore.fit(
#     X=dc_dfs['Ciências Humanas'][dc_preditoras['Ciências Humanas']],
#     y=dc_dfs['Ciências Humanas'][dc_target['Ciências Humanas']],
#     )

In [ ]:
df_humanas

In [ ]:
# ...existing code...
from sklearn.tree import DecisionTreeRegressor

# Select the dataframe and predictors for this subject
df_humanas = dc_dfs['Ciências Humanas'].copy()
cols_preditoras = dc_preditoras['Ciências Humanas']

# Identify categorical columns (object type) within the predictors
cat_cols = df_humanas[cols_preditoras].select_dtypes(include=['object']).columns

# Initialize and apply the encoder to categorical columns
encoder = OrdinalEncoder()
df_humanas[cat_cols] = encoder.fit_transform(df_humanas[cat_cols])

arvore = DecisionTreeRegressor(random_state=42)

arvore.fit(
    X=df_humanas[cols_preditoras],
    y=df_humanas[dc_target['Ciências Humanas'][0]],
    )

Vamos avaliar a correlação entre as variáveis preditoras, separando em grupos que fazem sentido.

In [ ]:
v_1 = ['fx_etaria', 'estado_civil', 'conclusao_ensino_medio', 'flag_treineiro']

for nome, df in dc_matriz_phik.items():

    print(f' {nome} ' .center(80, '='))

    df_posicao_phik = df.query(f'var_1 in {v_1} and var_2 in {v_1}').sort_values(['var_1', 'valor_phik'], ascending=False).reset_index(drop=True)

    print(df_posicao_phik)

In [ ]:
['fx_etaria',
 'sexo',
 'estado_civil',
 'cor_raca',
 'nacionalidade',
 'conclusao_ensino_medio',
 'flag_treineiro',
 'lingua_estrangeira',
 '01_escolaridade_pai',
 '02_escolaridade_mae',
 '03_ocupacao_pai',
 '04_ocupacao_mae',
 '05_qtde_moradores',
 '06_renda_familiar',
 '07_dias_trabalhador_domestico',
 '08_qtde_banheiro',
 '09_qtde_quarto',
 '10_qtde_carro',
 '11_qtde_motocicleta',
 '12_qtde_geladeira',
 '13_qtde_freezer',
 '14_qtde_maq_lavar_roupa',
 '15_qtde_maq_secar_roupa',
 '16_qtde_micro_ondas',
 '17_qtde_maq_lavar_louca',
 '18_flag_aspirador_po',
 '19_qtde_tv',
 '20_flag_aparelho_dvd',
 '21_flag_tv_assinatura',
 '22_qtde_celular',
 '23_flag_telefone_fixo',
 '24_qtde_computadores',
 '25_flag_internet']

In [ ]:
df_posicao_phik = pd.DataFrame(dc_preditoras['Ciências Humanas'], columns=['var'])

for nome, df in dc_dfs.items():

    target = dc_target[nome]

    df = dc_matriz_phik[nome].query('var_1 == @target and var_2 != @target').sort_values(by='valor_phik', ascending=False)[['var_2', 'valor_phik']]

    df['posicao_phik'] = range(1, len(df) + 1)

    df.drop(columns=['valor_phik'], inplace=True)

    df_posicao_phik = df_posicao_phik.merge(df.rename(columns={'var_2': 'var', 'posicao_phik': dc_nome_salvar[nome]}), on='var', how='left')

del target, df
gc.collect()

In [ ]:
df_posicao_phik.query('humanas <= 10 or natureza <= 10 or linguagem_codigo <= 10 or matematica <= 10 or redacao <= 10').reset_index(drop=True)

In [ ]:
for nome, df in dc_matriz_phik.items():

    target = dc_target[nome]

    df = df[(df['var_1'] == target) & (df['var_2'] != target)]

    print(f' Matriz Phik - {nome} '.center(100, '='))
    print()

    print(df.head(10))
    print()

In [ ]:
dc_matriz_phik['Ciências Natureza'].query('valor_phik != 1 and var_1 != "nota_ciencias_natureza" and var_2 != "nota_ciencias_natureza" and valor_phik > 0.7')

In [ ]:
for var in dc_preditoras['Ciências Humanas']:

    print(f' {var} '.center(60, '='))
    print()

    print(dc_dfs['Ciências Humanas'][var].unique())
    print(dc_dfs['Ciências Humanas'][var].isnull().sum())
    print()

In [ ]:
variavel = 'fx_etaria'

for k in dc_dfs.keys():

    df = dc_dfs[k]

    plt.figure(figsize = (15, 6))

    plt.hist(
        df[variavel],
        bins = 30,
        edgecolor = 'black',
        alpha = 0.7,
    )
    plt.xticks(sorted(df[variavel].unique()))
    plt.title(f"{k} - Distribuição por {variavel}")
    plt.show()

In [ ]:
# from phik import report

# for nome, df in dc_dfs.items():

#     print("=" * 50)
#     print()

#     print(f"Análise de correlação - {nome}")
#     print()

#     phik_overview = report.correlation_report(df[dc_preditoras[nome] + [dc_target[nome]]], bins = 10)

#     print(phik_overview)

#     print()